File: mobile/ios/packages/town-crier-presentation/Sources/Features/Onboarding/OnboardingViewModel.swift
import Combine
import TownCrierDomain

/// Steps in the first-launch onboarding flow.
public enum OnboardingStep: CaseIterable, Equatable, Sendable {
  case welcome
  case postcodeEntry
  case radiusPicker
  /// Custom-shape boundary drawing (GH#1031, tc-6he3x.10). Reached only via
  /// ``OnboardingViewModel/reconcileTierAfterCustomShapeUpgrade()`` after a
  /// successful in-wizard purchase from the radius step's custom-shape
  /// upsell — never part of the default welcome → postcode → radius →
  /// notification sequence.
  case boundaryDrawing
  case notificationPermission
}

/// Drives the onboarding flow: welcome → postcode entry → radius picker → notification permission → complete.
@MainActor
public final class OnboardingViewModel: ObservableObject, ErrorHandlingViewModel, BoundaryDrawingViewModel {
  @Published public private(set) var currentStep: OnboardingStep = .welcome
  @Published public var postcodeInput: String = ""
  @Published public private(set) var isLoading = false
  @Published public internal(set) var error: DomainError?
  private var validatedPostcode: Postcode?
  /// Set by ``prefill(name:coordinate:radiusMetres:)`` (GH#879 Phase 5): a
  /// device-local zone's name is free text the user chose (or the postcode
  /// it was seeded from), not necessarily a valid ``Postcode`` — so this
  /// carries the name straight through to the zone ``confirmRadius()``
  /// creates rather than round-tripping it through `Postcode` validation.
  private var prefillZoneName: String?
  @Published public private(set) var geocodedCoordinate: Coordinate?
  @Published public var selectedRadiusMetres: Double = 1000
  private var createdWatchZone: WatchZone?
  @Published public private(set) var isComplete = false

  /// The user's current subscription tier, injected at construction and kept
  /// fresh by ``AppCoordinator`` (it updates this in place when the live tier
  /// resolves, e.g. after an in-wizard purchase). It is `@Published` so the
  /// radius step can unlock the larger paid range reactively *without* the
  /// wizard being rebuilt — a `.id(tier)` rebuild would discard the in-progress
  /// postcode/geocode, which must survive the upgrade round-trip (tc-w3cb.3).
  @Published public internal(set) var subscriptionTier: SubscriptionTier

  /// Drives the in-wizard subscription paywall (tc-w3cb.3). Presented as a sheet
  /// *over* the wizard so the StateObject — and the in-progress postcode/geocode
  /// — survives the purchase round-trip.
  @Published public var isRadiusUpsellPresented = false

  /// Circle (the default) vs. custom polygon shape (GH#1031). Mirrors
  /// `WatchZoneEditorViewModel.shapeMode`: stays `.circle` for the whole
  /// default flow, and is only ever set to `.custom` by
  /// ``reconcileTierAfterCustomShapeUpgrade()`` when a mid-flow purchase
  /// unlocks ``currentStep`` == ``OnboardingStep/boundaryDrawing``. Tracked
  /// so ``goBack()`` can route `.notificationPermission` back to whichever
  /// of the radius or boundary step the zone was actually built from.
  @Published public private(set) var shapeMode: WatchZoneShapeMode = .circle

  /// The custom-shape ring being drawn, as an OPEN sequence of vertices in
  /// drop order — NOT closed. Mirrors
  /// `WatchZoneEditorViewModel.boundaryVertices`; see
  /// ``BoundaryDrawingViewModel`` for the shared surface `BoundaryDrawingMapView`
  /// needs to render it.
  @Published public private(set) var boundaryVertices: [Coordinate] = []

  /// Drives the in-wizard custom-shape upsell paywall (GH#1031, tc-6he3x.10),
  /// presented from the radius step alongside (but distinct from)
  /// ``isRadiusUpsellPresented``. Also a sheet *over* the wizard so the
  /// in-progress postcode/geocode survives the purchase round-trip.
  @Published public var isCustomShapeUpsellPresented = false

  var onComplete: ((WatchZone) -> Void)?

  /// Builds the paywall view-model for the in-wizard upsell sheet. Injected by
  /// ``AppCoordinator`` so the wizard stays decoupled from the composition root.
  /// Optional return so the view degrades gracefully if the factory is unset.
  var makeUpsellViewModel: (() -> SubscriptionViewModel?)?

  /// Invoked when the upsell sheet dismisses, so the coordinator can re-resolve
  /// the subscription tier and unlock the larger radius range live.
  var onUpgradeFlowCompleted: (() async -> Void)?

  private let geocoder: PostcodeGeocoder
  private let watchZoneRepository: WatchZoneRepository
  private let onboardingRepository: OnboardingRepository
  private let notificationService: NotificationService

  public init(
    geocoder: PostcodeGeocoder,
    watchZoneRepository: WatchZoneRepository,
    onboardingRepository: OnboardingRepository,
    notificationService: NotificationService,
    subscriptionTier: SubscriptionTier = .free
  ) {
    self.geocoder = geocoder
    self.watchZoneRepository = watchZoneRepository
    self.onboardingRepository = onboardingRepository
    self.notificationService = notificationService
    self.subscriptionTier = subscriptionTier
  }

  public func advance() {
    switch currentStep {
    case .welcome:
      currentStep = .postcodeEntry
    case .postcodeEntry:
      guard geocodedCoordinate != nil else { return }
      currentStep = .radiusPicker
    case .radiusPicker:
      currentStep = .notificationPermission
    case .boundaryDrawing:
      currentStep = .notificationPermission
    case .notificationPermission:
      break
    }
  }

  public func goBack() {
    switch currentStep {
    case .welcome:
      break
    case .postcodeEntry:
      currentStep = .welcome
    case .radiusPicker:
      currentStep = .postcodeEntry
    case .boundaryDrawing:
      // Stepping back out of drawing returns to the radius step as a plain
      // circle pick — both shapeMode and any in-progress vertices are
      // discarded, mirroring the editor's `selectShapeMode(.circle)`.
      shapeMode = .circle
      boundaryVertices = []
      currentStep = .radiusPicker
    case .notificationPermission:
      // Which step precedes notification depends on which shape the zone
      // was actually built from (tc-6he3x.10) — `shapeMode` is the source of
      // truth for that, since it's only ever `.custom` after a genuine
      // mid-flow tier change routed here via ``boundaryDrawing``.
      currentStep = shapeMode == .custom ? .boundaryDrawing : .radiusPicker
    }
  }

  /// Seeds the wizard's postcode/coordinate/radius from an already-resolved
  /// anonymous browse session and jumps straight to the radius step,
  /// skipping postcode entry a second time. Used for the anonymous browse
  /// post-signup handoff (GH#868 Phase 3.5, radius carried through in the
  /// Phase 3 refinement): a user who located themselves — and picked a live
  /// monitoring radius — before creating an account shouldn't be asked for
  /// either again. `radiusMetres` is clamped to `[100, maxRadiusMetres]` so a
  /// stale or legacy persisted value can never land outside the radius
  /// step's own slider bounds. Additive — the normal (non-prefilled) flow
  /// through `.postcodeEntry` -> ``submitPostcode()`` is unchanged for every
  /// other caller.
  public func prefill(postcode: Postcode, coordinate: Coordinate, radiusMetres: Double) {
    postcodeInput = postcode.value
    validatedPostcode = postcode
    geocodedCoordinate = coordinate
    selectedRadiusMetres = min(max(radiusMetres, 100), maxRadiusMetres)
    currentStep = .radiusPicker
  }

  /// Seeds the wizard's zone name/coordinate/radius from an already-created
  /// device-local zone (GH#879 Phase 5) and jumps straight to the radius
  /// step, mirroring ``prefill(postcode:coordinate:radiusMetres:)`` but for a
  /// zone whose name is arbitrary user text rather than a validated
  /// postcode. `radiusMetres` is clamped the same way.
  public func prefill(name: String, coordinate: Coordinate, radiusMetres: Double) {
    prefillZoneName = name
    postcodeInput = name
    validatedPostcode = nil
    geocodedCoordinate = coordinate
    selectedRadiusMetres = min(max(radiusMetres, 100), maxRadiusMetres)
    currentStep = .radiusPicker
  }

  public func submitPostcode() async {
    isLoading = true
    error = nil

    let postcode: Postcode
    do {
      postcode = try Postcode(postcodeInput)
    } catch {
      handleError(error)
      isLoading = false
      return
    }

    do {
      validatedPostcode = postcode
      geocodedCoordinate = try await geocoder.geocode(postcode)
      currentStep = .radiusPicker
    } catch {
      handleError(error)
    }

    isLoading = false
  }

  /// The maximum radius the user's tier permits, in metres. The radius slider
  /// is bounded at this so a free user cannot pick a zone larger than their
  /// tier allows — replacing the old discrete picker that wrongly offered 5 km
  /// to free accounts (cap 2 km). Shared source of truth with the editor via
  /// ``WatchZoneLimits`` (tc-w3cb.2).
  public var maxRadiusMetres: Double {
    WatchZoneLimits(tier: subscriptionTier).maxRadiusMetres
  }

  /// Whether the user's tier still has radius headroom to unlock — true for any
  /// tier below the top (Pro, 10 km). Drives the "Unlock larger zones" chip.
  public var canUnlockLargerRadius: Bool {
    subscriptionTier < .pro
  }

  /// Whether the user's tier delivers instant alerts (push and instant email).
  /// Free accounts receive only the weekly email digest, so the notification
  /// step adapts its copy — and shows a light upgrade nudge — accordingly
  /// (tc-w3cb.4). Same entitlement the editor's instant-alert toggles gate on.
  public var deliversInstantAlerts: Bool {
    EntitlementMap.hasEntitlement(.statusChangeAlerts, for: subscriptionTier)
  }

  /// Surfaces the in-wizard paywall when the user taps the unlock chip.
  public func requestLargerRadiusUpgrade() {
    isRadiusUpsellPresented = true
  }

  /// Called when the paywall sheet dismisses. Re-resolves the tier so a
  /// successful upgrade opens the larger radius range without rebuilding the
  /// wizard (which would discard the in-progress postcode/geocode).
  public func reconcileTierAfterUpgrade() async {
    await onUpgradeFlowCompleted?()
  }

  /// Whether to surface the "this zone may produce lots of notifications" callout
  /// (tc-1zb7). Triggered just above the free tier's 2 km cap so only paid tiers
  /// see it — see `LargeRadiusWarningView` for the threshold rationale.
  public var showsLargeRadiusWarning: Bool {
    selectedRadiusMetres >= LargeRadiusWarning.thresholdMetres
  }

  public func confirmRadius() {
    guard let coordinate = geocodedCoordinate else { return }
    do {
      let zone: WatchZone
      if let prefillZoneName {
        zone = try WatchZone(
          name: prefillZoneName, centre: coordinate, radiusMetres: selectedRadiusMetres)
      } else if let postcode = validatedPostcode {
        zone = try WatchZone(
          postcode: postcode, centre: coordinate, radiusMetres: selectedRadiusMetres)
      } else {
        return
      }
      createdWatchZone = zone
      currentStep = .notificationPermission
    } catch {
      self.error = .invalidWatchZoneRadius
    }
  }

  // MARK: - Custom-shape boundary drawing (GH#1031, tc-6he3x.10)

  /// Mirrors `WatchZoneBoundary`'s geometric minimum (3 distinct vertices
  /// are required to form a polygon).
  private static let minimumBoundaryVertexCount = 3

  /// Whether the user's tier may draw a custom-shape zone at all — gates
  /// showing the custom-shape upsell on the radius step. Mirrors
  /// `WatchZoneEditorViewModel.canDrawCustomShape` via the same
  /// ``WatchZoneLimits`` source of truth.
  public var canDrawCustomShape: Bool {
    WatchZoneLimits(tier: subscriptionTier).allowsCustomBoundary
  }

  /// Whether ``boundaryVertices`` currently has enough points to form a
  /// valid polygon (the domain layer's geometric minimum of 3). Drives the
  /// boundary step's "Continue" button.
  public var hasMinimumBoundaryVertices: Bool {
    boundaryVertices.count >= Self.minimumBoundaryVertexCount
  }

  /// Appends a new vertex at the end of the ring being drawn.
  public func addVertex(_ coordinate: Coordinate) {
    boundaryVertices.append(coordinate)
  }

  /// Moves the vertex at `index` to `coordinate` (e.g. after a drag). A
  /// stale or out-of-range index is a no-op rather than a crash.
  public func moveVertex(at index: Int, to coordinate: Coordinate) {
    guard boundaryVertices.indices.contains(index) else { return }
    boundaryVertices[index] = coordinate
  }

  /// Removes the vertex at `index`. A stale or out-of-range index is a
  /// no-op rather than a crash.
  public func removeVertex(at index: Int) {
    guard boundaryVertices.indices.contains(index) else { return }
    boundaryVertices.remove(at: index)
  }

  /// Removes the most recently added vertex. A no-op on an empty ring.
  public func undoLastVertex() {
    guard !boundaryVertices.isEmpty else { return }
    boundaryVertices.removeLast()
  }

  /// Validates the current vertices as a closed ring without saving —
  /// invoked when the user taps the first vertex to close the shape while
  /// drawing (``BoundaryDrawingMapView``). Surfaces problems (self-
  /// intersection, an out-of-UK vertex, a duplicate point) immediately via
  /// ``error``; below the minimum vertex count this is a silent no-op since
  /// there's nothing yet to validate. Mirrors
  /// `WatchZoneEditorViewModel.finishDrawing()`.
  public func finishDrawing() {
    guard hasMinimumBoundaryVertices else { return }
    do {
      _ = try WatchZoneBoundary(vertices: boundaryVertices)
      error = nil
    } catch {
      handleError(error)
    }
  }

  /// Surfaces the in-wizard custom-shape paywall when the user taps the
  /// upsell on the radius step.
  public func requestCustomShapeUpgrade() {
    isCustomShapeUpsellPresented = true
  }

  /// Re-enters the drawing step for a tier that already allows custom
  /// shapes — the way back in after ``goBack()`` leaves ``boundaryDrawing``,
  /// since the upsell card that reaches it hides once already entitled.
  public func selectCustomShape() {
    guard canDrawCustomShape, currentStep == .radiusPicker else { return }
    shapeMode = .custom
    currentStep = .boundaryDrawing
  }

  /// Called when the custom-shape paywall sheet dismisses. Re-resolves the
  /// tier via the same ``onUpgradeFlowCompleted`` hook as
  /// ``reconcileTierAfterUpgrade()``, then swaps `.radiusPicker` for
  /// `.boundaryDrawing` in place if the purchase unlocked custom shapes —
  /// preserving the already-entered postcode/geocode (tc-w3cb.3).
  public func reconcileTierAfterCustomShapeUpgrade() async {
    await onUpgradeFlowCompleted?()
    if currentStep == .radiusPicker, canDrawCustomShape {
      shapeMode = .custom
      currentStep = .boundaryDrawing
    }
  }

  /// Builds a custom-shape ``WatchZone`` from ``boundaryVertices`` and
  /// advances to the notification step — the ``boundaryDrawing`` step's
  /// counterpart to ``confirmRadius()``. The centre and radius are derived
  /// from the boundary's centroid and enclosing radius, exactly as
  /// `WatchZoneEditorViewModel.save()` does in `.custom` mode, so every
  /// existing circle-shaped read path (map centring, list rows) keeps
  /// working unchanged.
  public func confirmBoundary() {
    guard geocodedCoordinate != nil, hasMinimumBoundaryVertices else { return }
    do {
      let boundary = try WatchZoneBoundary(vertices: boundaryVertices)
      let centroid = boundary.centroid
      let centre = try Coordinate(latitude: centroid.latitude, longitude: centroid.longitude)
      let radius = boundary.enclosingRadiusMetres
      let zone: WatchZone
      if let prefillZoneName {
        zone = try WatchZone(
          name: prefillZoneName, centre: centre, radiusMetres: radius, boundary: boundary)
      } else if let postcode = validatedPostcode {
        zone = try WatchZone(
          postcode: postcode, centre: centre, radiusMetres: radius, boundary: boundary)
      } else {
        return
      }
      createdWatchZone = zone
      currentStep = .notificationPermission
    } catch {
      handleError(error)
    }
  }

  public func requestNotificationPermission() async {
    _ = try? await notificationService.requestPermission()
    await completeOnboarding()
  }

  public func skipNotifications() async {
    await completeOnboarding()
  }

  private func completeOnboarding() async {
    guard let zone = createdWatchZone else { return }
    try? await watchZoneRepository.save(zone)
    onboardingRepository.markOnboardingComplete()
    isComplete = true
    onComplete?(zone)
  }
}

File: mobile/android/presentation/src/main/kotlin/uk/towncrierapp/presentation/features/onboarding/OnboardingViewModel.kt
package uk.towncrierapp.presentation.features.onboarding

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import uk.towncrierapp.domain.auth.DomainError
import uk.towncrierapp.domain.onboarding.OnboardingRepository
import uk.towncrierapp.domain.subscriptions.Entitlement
import uk.towncrierapp.domain.subscriptions.FeatureGate
import uk.towncrierapp.domain.subscriptions.SubscriptionTier
import uk.towncrierapp.domain.watchzones.Postcode
import uk.towncrierapp.domain.watchzones.PostcodeGeocoder
import uk.towncrierapp.domain.watchzones.WatchZone
import uk.towncrierapp.domain.watchzones.WatchZoneId
import uk.towncrierapp.domain.watchzones.WatchZoneLimits
import uk.towncrierapp.domain.watchzones.WatchZoneRepository
import uk.towncrierapp.presentation.designsystem.components.LARGE_RADIUS_WARNING_THRESHOLD_METRES
import uk.towncrierapp.presentation.features.watchzones.PolygonDrawingState
import uk.towncrierapp.presentation.features.watchzones.ShapeForSave
import uk.towncrierapp.presentation.features.watchzones.WatchZoneShapeEvent
import uk.towncrierapp.presentation.features.watchzones.WatchZoneShapeMode
import uk.towncrierapp.presentation.features.watchzones.applyShapeEvent
import uk.towncrierapp.presentation.features.watchzones.resolveShapeForSave
import java.util.UUID

private val INSTANT_ALERT_ENTITLEMENT = Entitlement.STATUS_CHANGE_ALERTS

/**
 * Drives the onboarding wizard: postcode geocoding, tier-based radius
 * limits, tier-aware notification-permission copy, and the custom-shape
 * (polygon) drawing step (GH#1072 Phase 5, tc-v6fo0.5). Deliberately
 * constructed ONCE per wizard presentation - never re-created per step or on
 * a tier change - so [reconcileTierAfterUpgrade]/[reconcileTierAfterCustomShapeUpgrade]
 * can push a later tier into the SAME instance's state. That is what lets it
 * survive a future paywall dialog presented over the wizard (#783) without
 * losing the postcode/coordinate/radius already entered: the composition
 * root constructs this ViewModel once for the whole wizard route, not once
 * per step. Port of iOS `OnboardingViewModel`.
 */
@Suppress("TooManyFunctions")
// This wizard genuinely owns 16 independent, single-purpose interactions:
// the original 10 (step nav, postcode entry/lookup, radius edit/confirm,
// notification enable/skip/complete, reconcileTierAfterUpgrade) plus GH#1072
// Phase 5's 6 custom-shape actions (onShapeEvent - itself one dispatch point
// standing in for 5 shape/polygon actions, see WatchZoneShapeEvent -
// selectCustomShape, requestCustomShapeUpgrade, consumeNavigateToPaywall,
// reconcileTierAfterCustomShapeUpgrade, confirmBoundary). Splitting further
// means either merging semantically-distinct methods (breaks the
// cross-ViewModel consumeX()/confirmX() naming convention) or hiving off an
// unrelated class purely to dodge this count - neither improves readability
// over the function-count overage, same rationale as WatchZoneEditorViewModel.
public class OnboardingViewModel(
    private val postcodeGeocoder: PostcodeGeocoder,
    private val watchZoneRepository: WatchZoneRepository,
    private val onboardingRepository: OnboardingRepository,
    tier: SubscriptionTier,
    private val paywallAvailable: Boolean = false,
    private val enableDebugLogging: Boolean = false,
) : ViewModel() {
    private val _uiState = MutableStateFlow(initialState(tier, paywallAvailable))
    public val uiState: StateFlow<OnboardingUiState> = _uiState.asStateFlow()

    /**
     * Advances Welcome -> Postcode unconditionally, or Postcode -> Radius
     * once a postcode has been resolved. Radius -> NotificationPermission
     * goes through [confirmRadius] instead (and [OnboardingStep.BoundaryDrawing]
     * -> NotificationPermission through [confirmBoundary]), since both
     * transitions also build the in-memory zone.
     */
    public fun advance() {
        val next = nextStep(_uiState.value) ?: return
        _uiState.update { it.copy(step = next) }
    }

    /**
     * No-op on [OnboardingStep.Welcome] - the wizard is linear going forward
     * but not one-way. Stepping back out of [OnboardingStep.BoundaryDrawing]
     * discards [OnboardingUiState.shapeMode]/[OnboardingUiState.polygonVertices]
     * and returns to [OnboardingStep.Radius] as a plain circle pick, mirroring
     * the editor's mode toggle; stepping back from
     * [OnboardingStep.NotificationPermission] routes to whichever of Radius
     * or BoundaryDrawing the in-progress zone actually came from, using
     * [OnboardingUiState.shapeMode] as the source of truth (it is only ever
     * [WatchZoneShapeMode.CUSTOM] after a genuine [confirmBoundary] call).
     */
    public fun back() {
        val state = _uiState.value
        val previous = previousStep(state) ?: return
        _uiState.update {
            val withNewStep = it.copy(step = previous)
            if (state.step == OnboardingStep.BoundaryDrawing) withNewStep.discardCustomShapeDrawing() else withNewStep
        }
    }

    public fun updatePostcode(value: String) {
        _uiState.update { it.copy(postcodeInput = value) }
    }

    /** Validates the format locally first - a garbage postcode never spends a network call - before calling [postcodeGeocoder]. */
    public fun lookUpPostcode() {
        val raw = _uiState.value.postcodeInput
        val postcode = Postcode.parse(raw)
        if (postcode == null) {
            _uiState.update { it.copy(postcodeError = DomainError.GeocodingFailed(raw)) }
            return
        }
        viewModelScope.launch {
            _uiState.update { it.copy(isLookingUpPostcode = true, postcodeError = null) }
            try {
                val coordinate = postcodeGeocoder.geocode(postcode.value)
                _uiState.update {
                    it.copy(isLookingUpPostcode = false, resolvedPostcode = postcode, geocodedCoordinate = coordinate)
                }
            } catch (e: CancellationException) {
                throw e
            } catch (e: DomainError) {
                _uiState.update { it.copy(isLookingUpPostcode = false, postcodeError = e) }
            }
        }
    }

    public fun updateRadius(value: Float) {
        _uiState.update {
            it.copy(radiusMetres = value, showsLargeRadiusWarning = value >= LARGE_RADIUS_WARNING_THRESHOLD_METRES)
        }
    }

    /** Builds the [WatchZone] in memory only - nothing is saved until [completeOnboarding] - and advances to the notification-permission step. */
    public fun confirmRadius() {
        val state = _uiState.value
        val coordinate = state.geocodedCoordinate ?: return
        val limits = WatchZoneLimits(state.tier)
        val zone =
            WatchZone(
                id = WatchZoneId(UUID.randomUUID().toString()),
                name = state.resolvedPostcode?.value ?: state.postcodeInput.trim(),
                centre = coordinate,
                radiusMetres = limits.clampRadius(state.radiusMetres.toDouble()),
            )
        _uiState.update { it.copy(pendingZone = zone, step = OnboardingStep.NotificationPermission) }
    }

    // MARK: - Custom-shape boundary drawing (GH#1072 Phase 5, tc-v6fo0.5)

    /**
     * Routes every shape-mode/polygon-drawing interaction on
     * [OnboardingStep.BoundaryDrawing] through one dispatch point, same
     * shape as [uk.towncrierapp.presentation.features.watchzones.WatchZoneEditorViewModel.onShapeEvent]
     * - see [WatchZoneShapeEvent] for why a single method covers five
     * distinct user actions. The actual rules live in
     * [PolygonDrawingState.applyShapeEvent], shared with the editor rather
     * than duplicated.
     */
    public fun onShapeEvent(event: WatchZoneShapeEvent) {
        _uiState.update { state ->
            state.withPolygonDrawing(state.toPolygonDrawingState().applyShapeEvent(event, state.allowsCustomBoundary))
        }
    }

    /**
     * Re-entry point for a user whose tier already allows a custom shape (an
     * already-Personal/Pro user, or one who backed out of drawing and wants
     * back in) - mirrors iOS `selectCustomShape()`. Only takes effect while
     * sitting on [OnboardingStep.Radius] and entitled; a no-op from any other
     * step or tier, since [RadiusPickerStep] only ever shows this affordance
     * there.
     */
    public fun selectCustomShape() {
        _uiState.update { it.enterBoundaryDrawingIfEligible() }
    }

    /** Surfaces the future (#783) in-wizard custom-shape paywall when a free-tier user taps the upsell CTA on the radius step. */
    public fun requestCustomShapeUpgrade() {
        _uiState.update { it.copy(navigateToPaywall = true) }
    }

    /** Resets the one-shot route-to-paywall signal once the Route has acted on it. */
    public fun consumeNavigateToPaywall() {
        _uiState.update { it.copy(navigateToPaywall = false) }
    }

    /**
     * Pushes a tier raised by a (future, #783) purchase from the radius
     * step's custom-shape upsell into this SAME instance, via
     * [reconcileTierAfterUpgrade] - then, mirroring iOS
     * `reconcileTierAfterCustomShapeUpgrade()`, swaps [OnboardingStep.Radius]
     * for [OnboardingStep.BoundaryDrawing] ONLY when the user is actually
     * still sitting on the radius step and the new tier is now entitled. A
     * purchase completing after the user has already moved on (or that
     * somehow didn't unlock custom shapes) leaves the step untouched.
     */
    public fun reconcileTierAfterCustomShapeUpgrade(newTier: SubscriptionTier) {
        reconcileTierAfterUpgrade(newTier)
        _uiState.update { it.enterBoundaryDrawingIfEligible() }
    }

    /**
     * Builds the [WatchZone] from the drawn boundary and advances to the
     * notification-permission step - the [OnboardingStep.BoundaryDrawing]
     * counterpart to [confirmRadius]. Centre/radius are derived from the
     * boundary (centroid + enclosing radius), same as
     * [uk.towncrierapp.presentation.features.watchzones.WatchZoneEditorViewModel.save]'s
     * custom-shape path, so every circle-shaped read path (map centring,
     * list rows) keeps working unchanged. A no-op while the polygon isn't
     * closed yet (matches the disabled confirm button); surfaces
     * [OnboardingUiState.boundaryError] instead of advancing if the vertices
     * somehow fail [uk.towncrierapp.domain.watchzones.WatchZoneBoundary.of]
     * validation at this point.
     */
    public fun confirmBoundary() {
        val state = _uiState.value
        if (state.geocodedCoordinate == null) return
        when (val shape = state.toPolygonDrawingState().resolveShapeForSave()) {
            ShapeForSave.Circle, ShapeForSave.NotReady -> {
                Unit
            }

            ShapeForSave.Invalid -> {
                _uiState.update { it.copy(boundaryError = true) }
            }

            is ShapeForSave.Custom -> {
                val boundary = shape.boundary
                val zone =
                    WatchZone(
                        id = WatchZoneId(UUID.randomUUID().toString()),
                        name = state.resolvedPostcode?.value ?: state.postcodeInput.trim(),
                        centre = boundary.centroid,
                        radiusMetres = boundary.enclosingRadiusMetres,
                        boundary = boundary,
                    )
                _uiState.update { it.copy(pendingZone = zone, step = OnboardingStep.NotificationPermission) }
            }
        }
    }

    /**
     * Called when the user taps "Enable notifications". The actual OS
     * permission request is the Route's job (`rememberLauncherForActivityResult`
     * needs a composable scope) - this function's only job is to complete
     * the wizard regardless of the eventual grant/deny result, which is
     * intentionally never observed here.
     */
    public fun requestNotificationPermission() {
        completeOnboarding()
    }

    /** Free-tier's honest path: no OS permission was ever requested. */
    public fun skipNotifications() {
        completeOnboarding()
    }

    /**
     * Persists the in-memory zone best-effort - a failure is logged, not
     * blocking - sets the device latch, and signals completion. Matches iOS
     * `try?` semantics: the zones tab shows empty and the user can retry
     * from there rather than being trapped in the wizard.
     */
    public fun completeOnboarding() {
        viewModelScope.launch {
            _uiState.update { it.copy(isCompleting = true) }
            _uiState.value.pendingZone?.let { zone ->
                try {
                    watchZoneRepository.create(zone)
                } catch (e: CancellationException) {
                    throw e
                } catch (e: DomainError) {
                    log(enableDebugLogging) { "best-effort watch zone save failed at onboarding completion: $e" }
                }
            }
            onboardingRepository.setOnboardingComplete(true)
            _uiState.update { it.copy(isCompleting = false, isComplete = true) }
        }
    }

    /**
     * Pushes a tier raised by a (future, #783) paywall purchase into this
     * SAME instance - the caller re-resolves the tier after the purchase
     * flow and calls this rather than recreating the ViewModel, which is
     * what preserves postcode/coordinate/radius already entered. Also
     * refreshes [OnboardingUiState.allowsCustomBoundary] - see
     * [reconcileTierAfterCustomShapeUpgrade] for the mid-flow step swap this
     * alone does NOT perform.
     */
    public fun reconcileTierAfterUpgrade(newTier: SubscriptionTier) {
        _uiState.update { state ->
            val limits = WatchZoneLimits(newTier)
            val newMax = limits.maxRadiusMetres.toFloat()
            val clampedRadius = minOf(state.radiusMetres, newMax)
            state.copy(
                tier = newTier,
                maxRadiusMetres = newMax,
                radiusMetres = clampedRadius,
                showsLargeRadiusWarning = clampedRadius >= LARGE_RADIUS_WARNING_THRESHOLD_METRES,
                canUnlockLargerRadius = paywallAvailable && newTier < SubscriptionTier.PRO,
                hasInstantAlertEntitlement = FeatureGate(newTier).hasEntitlement(INSTANT_ALERT_ENTITLEMENT),
                allowsCustomBoundary = limits.allowsCustomBoundary,
            )
        }
    }
}

private fun nextStep(state: OnboardingUiState): OnboardingStep? =
    when (state.step) {
        OnboardingStep.Welcome -> OnboardingStep.Postcode
        OnboardingStep.Postcode -> if (state.geocodedCoordinate != null) OnboardingStep.Radius else null
        OnboardingStep.Radius, OnboardingStep.BoundaryDrawing, OnboardingStep.NotificationPermission -> null
    }

private fun previousStep(state: OnboardingUiState): OnboardingStep? =
    when (state.step) {
        OnboardingStep.Welcome -> {
            null
        }

        OnboardingStep.Postcode -> {
            OnboardingStep.Welcome
        }

        OnboardingStep.Radius -> {
            OnboardingStep.Postcode
        }

        OnboardingStep.BoundaryDrawing -> {
            OnboardingStep.Radius
        }

        OnboardingStep.NotificationPermission -> {
            if (state.shapeMode == WatchZoneShapeMode.CUSTOM) OnboardingStep.BoundaryDrawing else OnboardingStep.Radius
        }
    }

private fun initialState(
    tier: SubscriptionTier,
    paywallAvailable: Boolean,
): OnboardingUiState {
    val limits = WatchZoneLimits(tier)
    return OnboardingUiState(
        tier = tier,
        maxRadiusMetres = limits.maxRadiusMetres.toFloat(),
        canUnlockLargerRadius = paywallAvailable && tier < SubscriptionTier.PRO,
        hasInstantAlertEntitlement = FeatureGate(tier).hasEntitlement(INSTANT_ALERT_ENTITLEMENT),
        allowsCustomBoundary = limits.allowsCustomBoundary,
    )
}

// MARK: - Custom-shape boundary drawing helpers (GH#1072 Phase 5, tc-v6fo0.5)
//
// Top-level, not class members, for the same TooManyFunctions/complexity
// budget reason as WatchZoneEditorViewModel's equivalent private helpers.

private fun OnboardingUiState.toPolygonDrawingState(): PolygonDrawingState =
    PolygonDrawingState(
        shapeMode = shapeMode,
        polygonVertices = polygonVertices,
        isPolygonClosed = isPolygonClosed,
        canClosePolygon = canClosePolygon,
        boundaryError = boundaryError,
    )

private fun OnboardingUiState.withPolygonDrawing(drawing: PolygonDrawingState): OnboardingUiState =
    copy(
        shapeMode = drawing.shapeMode,
        polygonVertices = drawing.polygonVertices,
        isPolygonClosed = drawing.isPolygonClosed,
        canClosePolygon = drawing.canClosePolygon,
        boundaryError = drawing.boundaryError,
    )

/** Discards any in-progress drawing and reverts to [WatchZoneShapeMode.CIRCLE] - what stepping back out of [OnboardingStep.BoundaryDrawing] does. */
private fun OnboardingUiState.discardCustomShapeDrawing(): OnboardingUiState = withPolygonDrawing(PolygonDrawingState())

/**
 * Swaps [OnboardingStep.Radius] for [OnboardingStep.BoundaryDrawing] when -
 * and only when - currently on the radius step and [OnboardingUiState.allowsCustomBoundary]
 * is true. Shared by [OnboardingViewModel.selectCustomShape] and
 * [OnboardingViewModel.reconcileTierAfterCustomShapeUpgrade].
 */
private fun OnboardingUiState.enterBoundaryDrawingIfEligible(): OnboardingUiState =
    if (step == OnboardingStep.Radius && allowsCustomBoundary) {
        copy(shapeMode = WatchZoneShapeMode.CUSTOM, step = OnboardingStep.BoundaryDrawing)
    } else {
        this
    }

private inline fun log(
    enableDebugLogging: Boolean,
    message: () -> String,
) {
    // Guarded so plain JVM unit tests (enableDebugLogging = false by
    // default) never touch android.util.Log - mirrors ApiClient's `log`.
    if (enableDebugLogging) Log.w("OnboardingViewModel", message())
}
